### PR TITLE
ggml : fix padding in timestep embedding kernels

### PR DESCRIPTION
**Pull request by danbev**
This commit applies the same changes as were applied in #15917 to the cuda, metal, opencl, sycl, and vulkan backends.

Refs: https://github.com/ggml-org/llama.cpp/pull/15917

### CODE REVIEW

**Review by ggerganov — APPROVED** [review]
I think we should add a comment to the declaration of this function to clarify that it will pad the tensor if it has odd number of dimensions. So here:

https://github.com/ggml-org/llama.cpp/blob/28b5f190ef1dbea5edf82dbc8b4407b721fadd13/ggml/include/ggml.h#L2123-L2130

Change to:

```c
// return: [N, PAD(dim, 2)]
```

Btw, not sure what was the reasoning for this padding logic. It does not seem to be present in the reference Python implementation. Maybe @leejet can clarify?

**Comment by danbev** [comment]
I forgot to add `[no ci]` so I manually cancelled the workflows, hence the "Some checks haven't completed yet".

**Comment by leejet** [comment]
> I think we should add a comment to the declaration of this function to clarify that it will pad the tensor if it has odd number of dimensions. So here:
> 
> https://github.com/ggml-org/llama.cpp/blob/28b5f190ef1dbea5edf82dbc8b4407b721fadd13/ggml/include/ggml.h#L2123-L2130
> 
> Change to:
> 
> ```c
> // return: [N, PAD(dim, 2)]
> ```
> 
> Btw, not sure what was the reasoning for this padding logic. It does not seem to be present in the reference Python implementation. Maybe @leejet can clarify?

It seems that padding is incorrect behavior. I don’t really remember why I added it in the first place—it was too long ago. Given the current situation, I suggest removing the padding.

**Comment by danbev** [comment]
@leejet Thanks for the feedback! 
I'll take a closer look at removing the padding next week.

**Comment by danbev** [comment]
@ggerganov @leejet I've taken a closer look at this and I think that the padding might be required after all.

Looking at the reference python [implementation](https://github.com/CompVis/stable-diffusion/blob/main/ldm/modules/diffusionmodules/util.py#L151) we have the following:                                                                  
```python                                                                       
def timestep_embedding(timesteps, dim, max_period=10000, repeat_only=False):    
    if not repeat_only:                                                            
        half = dim // 2                                                         
        freqs = torch.exp(                                                      
            -math.log(max_period) * torch.arange(start=0, end=half, dtype=torch.float32) / half
        ).to(device=timesteps.device)                                           
        args = timesteps[:, None].float() * freqs[None]                         
        embedding = torch.cat([torch.cos(args), torch.sin(args)], dim=-1)       
        if dim % 2:                                                             
            embedding = torch.cat([embedding, torch.zeros_like(embedding[:, :1])], dim=-1)
    else:                                                                       
        embedding = repeat(timesteps, 'b -> b d', d=dim)                        
    return embedding                                                            
```
Specifically these lines:                                                                   
```python                                                                       
        if dim % 2:                                                             
            embedding = torch.cat([embedding, torch.zeros_like(embedding[:, :1])], dim=-1)
``` 
`if dim % 2` checks if the dimension is odd, and if so it adds an extra column of zeros to make the final embedding have the correct dimension. 

This padding is necessary because sinusoidal embeddings produce pairs of cosine/sine values. For dim=15, you get 7 complete pairs (14 values) plus 1 padding zero to reach the target dimension. Without this padding, the GGML implementation leaves uninitialized memory that can contain garbage values, which is what we observed in the test failures. I believe the padding in GGML is required to match this reference behavior.

**Comment by ggerganov** [comment]
@danbev I think the problem is that for input `dim=15` the Python code produces output dimension of `15`, while `ggml` produces `16`. At least this is what I understand from reading the implementation in `ggml.c`:

https://github.com/ggml-org/llama.cpp/blob/261e6a20ffdb79c4875e674b4f6b514bc73cff8f/ggml/src/ggml.c#L4919-L4932

**Comment by danbev** [comment]
@ggerganov Oh right, I misunderstood the comment about removing the padding to mean in the kernels. It indeed looks like adding an extra dimension here would not be needed. I'll update this and test it.

**Review by ggerganov — APPROVED** [review]
